application/client.core/src/app/environment/components/sidebar/shell/input/component.ts

```diff
@@ -0,0 +1,209 @@
+            this._ng_autoComRef.closePanel();
+        }, 250);
+        this._selectedTextOnInputClick = false;
+    }
+
+    public _ng_onAutocompletePanelOpen() {
+        if (this._ng_autoComRef === undefined || this._ng_autoComRef === null) {
+            return;
+        }
+        this._ng_autoComRef.updatePosition();
+    }
+
+    public _ng_onRecentSelected(event: MatAutocompleteSelectedEvent) {
+        this._ng_inputCtrl.setValue(event.option.viewValue);
+        if (!this._selectedTextOnInputClick) {
+            this._ng_onKeyUpRequestInput();
+        }
+    }
+
+    public _ng_getSafeHTML(input: string): SafeHtml {
+        return this._sanitizer.bypassSecurityTrustHtml(input);
+    }
+
+    public _ng_clearRecent() {
+        this._ng_autoComRef.closePanel();
+        this._ng_inputCtrl.updateValueAndValidity();
+        ElectronIpcService.request(new IPCMessages.ShellRecentCommandsClearRequest(), IPCMessages.ShellRecentCommandsClearResponse).then((response: IPCMessages.ShellRecentCommandsClearResponse) => {
+            this._loadRecentCommands();
+            if (response.error) {
+                return this._logger.error(`Fail to reset recent commands due error: ${response.error}`);
+            }
+        }).catch((error: Error) => {
+            return this._logger.error(`Fail send request to reset recent commands due error: ${error.message}`);
+        });
+    }
+
+    private _runCommand(command: string) {
+        ElectronIpcService.request(new IPCMessages.ShellProcessRunRequest({ session: ShellService.session.getGuid(), command: command }), IPCMessages.ShellProcessRunResponse).then((response: IPCMessages.ShellProcessRunResponse) => {
+            if (response.error !== undefined) {
+                this._logger.error(`Failed to run command due Error: ${response.error}`);
+                this._notificationsService.add({
+                    caption: 'Fail to run command',
+                    message: `Fail to run command ${command} due error: ${response.error}`
+                });
+            }
+        }).catch((error: Error) => {
+            this._logger.error(`Failed to run command due Error: ${error.message}`);
+            this._notificationsService.add({
+                caption: 'Fail to run command',
+                message: `Fail to run command ${command} due error: ${error.message}`
+            });
+        });
+    }
+
+    private _addCommand(command: string) {
+        ElectronIpcService.request(new IPCMessages.ShellRecentCommandAddRequest({ command: command }), IPCMessages.ShellRecentCommandAddResponse).then((response: IPCMessages.ShellRecentCommandAddResponse) => {
+            if (response.error !== undefined) {
+                this._logger.error(`Failed to add command to recent commands due Error: ${response.error}`);
+            }
+        }).catch((error: Error) => {
+            this._logger.error(`Failed to add command to recent commands due Error: ${error.message}`);
+        });
+    }
```

Comment:
I think it isn't necessary to send a command from FE to add something into recent. Because on BE we know - was the command successful or not, we can save the command into recent without a request from FE. I mean as soon as the command was successful, we just add it into recent on BE and do not until FE will ask to do it.